ivy/src/main/scala/sbt/librarymanagement/ivy/IvyDependencyResolution.scala

```diff
@@ -21,6 +21,16 @@ class IvyDependencyResolution private[sbt] (val ivySbt: IvySbt)
   ): Either[UnresolvedWarning, UpdateReport] =
     IvyActions.updateEither(toModule(module), configuration, uwconfig, log)
 
+  override def withAddedResolvers(
+      resolvers: Seq[Resolver],
+      log: Logger
+  ): IvyDependencyResolution = {
+    ivySbt.withIvy(log) { ivy =>
```

Comment:
I guess currently Resolver passing is done somewhat outside of LM API - https://github.com/sbt/sbt/blob/v1.9.4/main/src/main/scala/sbt/Defaults.scala#L4022-L4038. I've been reluctant to give too much support to Resolvers because beyond the basic URL layout for HTTPS there's no consensus among the LM engine implementations on how resolvers are used. For example, someone has implemented Amazon S3 resolver for Ivy, that can't be passed into Coursier.

In general, we don't have LM API for LM engine construction beyond `IvyDependencyResolution.apply` function, so the best we can do is maybe start there and implement another `apply` that takes additional resolvers, and also make sure we can implement one for Coursier as well.

Reply:
> I guess currently Resolver passing is done somewhat outside of LM API - https://github.com/sbt/sbt/blob/v1.9.4/main/src/main/scala/sbt/Defaults.scala#L4022-L4038. 

Thanks for the tip, I managed to figure out a more principled way to add the extra resolvers (`configuration` is either an `InlineIvyConfiguration` or an `ExternalIvyConfiguration` both of these implementations provided an easy way to add extra resolvers).

I just pushed a commit with the changes

> I've been reluctant to give too much support to Resolvers because beyond the basic URL layout for HTTPS there's no consensus among the LM engine implementations on how resolvers are used. For example, someone has implemented Amazon S3 resolver for Ivy, that can't be passed into Coursier.

I guess I am hitting contention here, which is treating `DependencyResolution` (and by proxy LM Engine) as a first class citizen vs Ivy Support which I have the view that it should eventually be deprecated but considering your response maybe this is a case of "lets not go there".

My immediate (and most likely non-educated) answer to this would be that we should extend the `DependencyResolution` interface to cover our use cases (just as is being done in this PR) and going forward if someone wants to, for example implement an S3 resolver (I assume you are talking about things like https://github.com/ohnosequences/ivy-s3-resolver) we would extend `sbt.librarymanagement.Resolver` to cover cases like this.

But then we circle back to your point in that there is no consensus on how to extend `sbt.librarymanagement.Resolver` (which is the point you are bringing up in the first place) however I only see a very minor tangental link between this point and my PR. More specifically I am missing how this would effect (either positively or negatively, if at all) any future improvement/design/suggestion to add cross resolver support for Coursier/Ivy.

> In general, we don't have LM API for LM engine construction beyond `IvyDependencyResolution.apply` function, so the best we can do is maybe start there and implement another `apply` that takes additional resolvers, and also make sure we can implement one for Coursier as well.

As said previously, I just updated the PR so the builder pattern works as I originally wanted it to, should I also add the `Resolver` to apply? Strictly speaking for the problem I am solving this is not needed so my initial inclination is to leave it to a future PR which may be more appropriate given that we are already discussing open design points.

To expand, the whole intent of the PR is for users to *add* resolvers ontop of the existing list of resolvers, putting that in `apply` would be a bit confusing considering that the `apply` would typically contain the initial set of resolvers (which is already provided by `configuration`) or am I missing something?

Reply:
@eed3si9n Now that sbt 2.0 is going forwards, what are your thoughts on this

Reply:
Also pinging @jtjeferreira since I think you seem to be a maintainer of this project now

Reply:
I am out of depth here, but maybe it would be worth to do this when https://github.com/sbt/sbt/pull/7739 is finished,  so we can do all the necessary changes in a single PR...

Reply:
> I am out of depth here, but maybe it would be worth to do this when [sbt/sbt#7739](https://github.com/sbt/sbt/pull/7739) is finished, so we can do all the necessary changes in a single PR...

Sure!

@adpi2 Maybe you want to chime in here as well

Reply:
I have to think a bit more on the technical details of the proposal, but one thing I can tell you is that we've removed `useCoursier` flag from sbt 2.x, which means that out-of-box everyone would be assumed to use Coursier. Maybe someone can write a plugin to revive Ivy as LM engine, but I'm ever more determined to make Coursier the default.

So if we're thinking about changing the LM API, my primary concern would be how it would align with the semantics of Coursier impl.

Reply:
Regarding making coursier a default, that is definitely a welcome change and I fully support it. If thats the case then actually the concept of this should be trivialized no, we just need to add the equivalent implementation methods to coursier integration in sbt and then it should work? Not sure what you had in mind regarding technical proposal, but I remember doing this locally (i.e. also modifying sbt's internal bridge to coursier for `DependencyResolution`) and it worked as expected. iirc when I implemented this, Coursier's resolvers was just an immutable collection and add another while returning a new instance of `DependencyResolution` worked as expected.

Judging from https://github.com/sbt/sbt/pull/7753, it appears that librarymanagement is getting inlined into sbt proper rather than its own github repo/module, if thats the case then it should be quite easy to implement this as we don't have a chicken and egg problem.